scanner/decisions/0005-scannerv4-read-cvss-data.md

```diff
@@ -0,0 +1,36 @@
+# 0005 - Adapt Scanner V4 NVD CVSS data updating pipeline to integrate with NVD CVE Api
+
+- **Author(s):** Yi Li <yli3@redhat.com>
+- **Created:** [2023-10-10]
+
+## Status
+
+Accepted
+
+## Context
+
+The National Vulnerability Database (NVD) has announced the retirement of its JSON feeds by December 15, 2023. This creates the need to change our existing CVSS data updater pipeline for Scanner V4, which currently relies on these JSON feeds. 
+
+The NVD JSON feed offers bundled data on a yearly basis, spanning from 2002 to the present year. A benefit is that only two requests are needed in the workflow for each year (one for the meta file and one for the data bundle). However, a limitation is the absence of filtering capabilities, such as by CVSS score or modification date. In contrast, the NVD API allows data retrieval through specific HTTP URL parameters, such as publish date, modified date, or CVSS score, enabling more precise data acquisition.
+
+## Decision
+
+The updated NVD CVSS GitHub Workflow (CI) has transitioned to sourcing data from the NVD CVE API, rather than the NVD Json feeds, ensuring the timely update of the NVD CVSS data bundle in Google Storage. To fetch data, the workflow downloads based on both the start index and CVSS v3 severity. For instance, a command such as curl "https://services.nvd.nist.gov/rest/json/cves/2.0?cvssV3Severity=CRITICAL&startIndex=0" > critical-1.json would be used. For every severity level, the workflow iteratively exhaust the startIndex to capture all associated data. And there are four severity levels in total. After all json files downloaded, they will be compress to one data bundle as one single zip file.
+
+We've chosen to base our API requests on CVSS V3 severity to ensure all data integrated into Claircore possesses valid CVSS V3 metrics. Consequently, retrieving data without CVSS V3 information becomes irrelevant.
+
+The CVSS Updater in Central will be modified to download zip file (generated by CI mentioned above) in the Google bucket. This eliminates the need for JSON parsing in the updater. ScannerV4 will pull the zip bundle from Central and populate the Matcher DB with the enrichment data.
+
+The CVSS Updater in Central will attempt to download the data bundle from the Google bucket a maximum of five times. Should all five attempts be unsuccessful, the updater will pause its efforts and initiate a fresh download attempt after a 4-hour interval. After five unsuccessful download attempts, a failure notification will be swiftly sent to the Scanner on-call Slack channel.
```

Comment:
Maybe too much of an implementation detail for this, but will we be using some sort of exponential backoff mechanism? Will we care about the status code or any code that is not 200 nor 304 will trigger a retry?

Also, live Centrals will message our Slack channel? I don't think there is much we can do about this outside of ACSCS. I don't think we should be notified by this, as there can be many reasons downloads fail. For example, if a customer does not set up the firewall properly, there is nothing we can do, and we would not want to be spammed with errors.

We could always test out the downloads ourselves in periodic CI for sanity checks. Similar to what we have now

Reply:
I agree. "A failure notification will be swiftly sent to the Scanner on-call Slack channel." was too much. I will remove it.

Reply:
> Maybe too much of an implementation detail for this, but will we be using some sort of exponential backoff mechanism? Will we care about the status code or any code that is not 200 nor 304 will trigger a retry?
> 
> Also, live Centrals will message our Slack channel? I don't think there is much we can do about this outside of ACSCS. I don't think we should be notified by this, as there can be many reasons downloads fail. For example, if a customer does not set up the firewall properly, there is nothing we can do, and we would not want to be spammed with errors.
> 
> We could always test out the downloads ourselves in periodic CI for sanity checks. Similar to what we have now

We can use exponential backoff mechanism to retry when the error code is 429 in case there's a rate limiting. Also we can retry 5xx Server Errors, this might help with clients with proxy servers.

I agree. "A failure notification will be swiftly sent to the Scanner on-call Slack channel." was too much. I will remove it.